src/vm_state/helpers.rs

```diff
@@ -205,6 +205,18 @@ impl<
             .decommittment_processor
             .prepare_to_decommit(monotonic_cycle_counter, partial_query)?;
 
+        if query.is_fresh {
+            assert!(
+                candidate_page == query.memory_page,
```

Comment:
and what happens when we execute 2 DIFFERENT decommits in the same memory page?

Reply:
it should also never happen, but it is hard to enforce on zk_evm level, much easier to put such asserts in zksync-era

Reply:
here I disagree. spreading such hidden assumptions is very risky. I'd rather have them all in this crate.

Reply:
zk_evm does not know anything about whether a code hash has been decommitted before or not. This is why code decommitter it abstracted away as an interface. To protect against such case we would have to maintain a map of code_hash -> mem_page in the `VmState` itself and change the interface of decommiter to just `give_preimage(h256) -> vec<u8>` function and let the zk_evm crate handle all the rest.

I am not against it, but it will introduce quite some changes, I think too many at this stage.